xla/backends/gpu/runtime/command_buffer_cmd.cc

```diff
@@ -867,260 +838,489 @@ CommandBufferCmd::BufferUseVector Memset32Cmd::buffers() {
       num_iterations_(num_iterations),
       loop_counter_(loop_counter),
-      body_commands_(std::move(body_commands)) {}
+      body_commands_(std::move(body_commands)) {
+  body_and_predict_commands_ = std::make_unique<CommandBufferCmdSequence>();
+  body_and_predict_commands_->Append(
+      std::make_unique<ChildCmd>(body_commands_->Clone()));
+  body_and_predict_commands_->Append(std::make_unique<SetForConditionCmd>(
+      &cond_handle_, loop_counter_, num_iterations_));
+}
 
 absl::Status ForCmd::Initialize(const Thunk::InitializeParams& params,
                                 StateManager& state) {
-  return body_commands_.Initialize(params, state);
+  return body_and_predict_commands_->Initialize(params, state);
 }
 
 absl::Status ForCmd::Record(const Thunk::ExecuteParams& execute_params,
                             const RecordParams& record_params,
-                            se::CommandBuffer* command_buffer) {
+                            se::CommandBuffer* command_buffer, bool create) {
   se::DeviceMemoryBase loop_counter =
       execute_params.buffer_allocations->GetDeviceAddress(loop_counter_);
 
-  ExecutionScopeId execution_scope_id = GetExecutionScope(record_params);
   VLOG(5) << "ForCmd: num_iterations=" << num_iterations_
-          << "; body_commands=" << body_commands_.size()
-          << "; execution_scope_id=" << execution_scope_id.value();
-  VLOG(5) << "  loop_counter: " << loop_counter_ << " ("
+          << "; body_commands=" << body_and_predict_commands_->size()
+          << "  loop_counter: " << loop_counter_ << " ("
           << loop_counter.opaque() << ")";
 
-  return command_buffer->For(
-      execution_scope_id, num_iterations_,
-      se::DeviceMemory<int32_t>(loop_counter),
-      CreateBuilder(&body_commands_, &execute_params, &record_params));
+  if (create) {
+    TF_ASSIGN_OR_RETURN(initialize_counter_node_,
+                        command_buffer->CreateMemsetNode(
+                            ToDependentNodes(), loop_counter, uint32_t(0),
+                            /*num_elements=*/1));
+
+    TF_ASSIGN_OR_RETURN(cond_handle_,
+                        command_buffer->CreateConditionalHandle());
+    TF_ASSIGN_OR_RETURN(
+        set_cond_handle_node_,
+        command_buffer->CreateForConditionNode(
+            Dependencies{initialize_counter_node_}, cond_handle_,
+            se::DeviceMemory<int32_t>(loop_counter), num_iterations_));
+    TF_ASSIGN_OR_RETURN(auto cond_node_result,
+                        command_buffer->CreateConditionalNode(
+                            Dependencies{set_cond_handle_node_}, cond_handle_,
+                            se::CommandBuffer::ConditionType::kWhile));
+    body_command_buffer_ = std::move(cond_node_result.command_buffer);
+    cond_node_ = cond_node_result.node_handle;
+  } else {
+    TF_RETURN_IF_ERROR(command_buffer->UpdateMemsetNode(
+        initialize_counter_node_, loop_counter, uint32_t(0),
```

Comment:
Ditto. Can you use `static_cast<uint32_t>` instead?

Reply:
Fixed

Reply:
@golechwierowicz Could you check again why google internal tests failed?

Reply:
```
module compiler/xla/stream_executor:command_buffer does not directly depend on a module exporting 'third_party/absl/container/flat_hash_set.h'
```

it seems like just the dependency on absl::flat_hash_set is missing in BUILD files.

Reply:
Thanks, just pushed a fix commit.

Reply:
@golechwierowicz Could you help check again?

Reply:
There's way more errors like this now. Let me handle this internally.

Reply:
Thanks you. @golechwierowicz

Reply:
@golechwierowicz Just found an issue in this PR, it may break google internal tests, just submit the fix.

Reply:
@ezhulenev said he will take over